Add updated information on domain parameter to OAuth docs (#18943)
DataDog/documentation: content/en/developers/authorization/oauth2_in_datadog.md

3. Once a user signs in, redirect them to the [OAuth2 Authorization endpoint][6] with the appropriate URL parameters, which includes the added `code_challenge` parameter generated by your application.

   To learn how to derive the `code_challenge` parameter, see the [PKCE](#authorization-code-grant-flow-with-pkce) section. Your application must save `code_verifier` for the token request in Step 5.

   - To build the URL for this GET request, use the `domain` query parameter that is provided on the redirect to your `onboarding_url`. 
   - This parameter is only provided if the user initiates authorization from the Datadog integration tile. See the [Initiate authorization from a third-party location](#Initiate-authorization-from-a-third-party-location) section for more options if the user chooses to initiate authorization externally.  
   - The `domain` query parameter provides the [Datadog site][17] that the authorizing user is in, and is required to construct the URL for this GET request to the Authorize endpoint: `https://api.<domain>/oauth2/v1/authorize?...`.

> <div class="alert alert-info">The `domain` parameter was introduced in July 2023, and previous OAuth clients use the `site` parameter for this step. While the `site` parameter is still supported, Datadog recommends using the `domain` parameter for the `Authorize` endpoint. </div>

Reviewer: Based on our previous conversation, should we say "**Store this value in a secure database or location for later use with additional API endpoints**. "

Author: Yeah, I can open a PR to add that

   To learn how to derive the `code_challenge` parameter, see the [PKCE](#authorization-code-grant-flow-with-pkce) section. Your application must save `code_verifier` for the token request in Step 5.

   - To build the URL for this GET request, use the `domain` query parameter that is provided on the redirect to your `onboarding_url`. 
   - This parameter is only provided if the user initiates authorization from the Datadog integration tile. See the [Initiate authorization from a third-party location](#Initiate-authorization-from-a-third-party-location) section for more options if the user chooses to initiate authorization externally.  
   - The `domain` query parameter provides the [Datadog site][17] that the authorizing user is in, and is required to construct the URL for this GET request to the Authorize endpoint: `https://api.<domain>/oauth2/v1/authorize?...`.

> <div class="alert alert-info">The `domain` parameter was introduced in July 2023, and previous OAuth clients use the `site` parameter. While the `site` parameter is supported, Datadog recommends using the `domain` parameter. </div>

Reviewer: ```suggestion
<div class="alert alert-info">The `domain` parameter was introduced in July 2023, and previous OAuth clients use the `site` parameter. While the `site` parameter is supported, Datadog recommends using the `domain` parameter. </div>
```
Not sure if we want to call out that it's partially supported since it won't work for endpoints like `logs`

Author: I don't think we need to call this out specifically

   To learn how to derive the `code_challenge` parameter, see the [PKCE](#authorization-code-grant-flow-with-pkce) section. Your application must save `code_verifier` for the token request in Step 5.

>    - In order to build the URL for this POST request, use the `domain` query parameter that is provided on the redirect to your `onboarding_url`. 

Author: @bgoldberg122 isn't this a GET request?

Reviewer: I think it should be GET. According to the docs that we link [here](https://docs.datadoghq.com/developers/authorization/oauth2_endpoints/?tab=authorizationendpoints#request-authorization-from-a-user)

Author: yeah - updating

>    - In order to build the URL for this POST request, use the `domain` query parameter that is provided on the redirect to your `onboarding_url`. 

Author: ```suggestion
   - To build the URL for this GET request, use the `domain` query parameter that is provided on the redirect to your `onboarding_url`. 
```

>    - In order to build the URL for this POST request, use the `domain` query parameter that is provided on the redirect to your `onboarding_url`. 

Reviewer: ```suggestion
   - In order to build the URL for this GET request, use the `domain` query parameter that is provided on the redirect to your `onboarding_url`. 
```

   - This parameter is only provided if the user initiates authorization from the Datadog integration tile. See the [Initiate authorization from a third-party location](#Initiate-authorization-from-a-third-party-location) section for more options if the user chooses to initiate authorization externally.  

>    - The `domain` query parameter provides the [Datadog site][17] that the authorizing user is in, and will be needed to construct the URL for this POST request to the Authorize endpoint: 

Reviewer: ```suggestion
   - The `domain` query parameter provides the [Datadog site][17] that the authorizing user is in, and will be needed to construct the URL for this GET request to the Authorize endpoint: 
```

>        - For users authorizing from the US1, Fed1, and EU1 Datadog sites, the URL will be: `https://app.<domain>/oauth2/v1/authorize?...`

Author: ```suggestion
       - For users authorizing from the US1, Fed1, and EU1 Datadog sites, the URL is: `https://app.<domain>/oauth2/v1/authorize?...`
```

>        - For users authorizing from any other Datadog site, the URL will be: `https://<domain>/oauth2/v1/authorize?...`

Author: ```suggestion
       - For users authorizing from any other Datadog site, the URL is: `https://<domain>/oauth2/v1/authorize?...`
```

> <div class="alert alert-info">The `domain` parameter is new as of July 2023. OAuth clients created before then use the `site` parameter instead. The `site` parameter will continue to be supported, but we encourage use of the `domain` parameter to make API calls more seamless. </div>

Author: ```suggestion
<div class="alert alert-info">The `domain` parameter is new as of July 2023. OAuth clients created before then use the `site` parameter instead. The `site` parameter will continue to be supported, but the `domain` parameter is recommended moving forward. </div>
```

4. Once a user clicks **Authorize**, Datadog makes a POST request to the authorize endpoint. The user is redirected to the `redirect_uri` that you provided when setting up the OAuth Client with the authorization `code` parameter in the query component.

5. From the `redirect_uri`, make a POST request to the [Datadog token endpoint][10] that includes the authorization code from Step 4, the `code_verifier` from Step 3, your OAuth client ID, and client secret.

>    - In order to build the URL for this post request, use the `domain` query parameter that is provided on the redirect to your `redirect_uri`. 

Author: ```suggestion
   - To build the URL for this post request, use the `domain` query parameter that is provided on the redirect to your `redirect_uri`. 
```

>    - In order to build the URL for this post request, use the `domain` query parameter that is provided on the redirect to your `redirect_uri`. 

Reviewer: We don't control this flow -- I think the param here might still be `site`

Author: Oh i see - for the token endpoint they'd still use site, but for the authorization endpoint they'd use domain?

Reviewer: Yeah, I'm not 100% sure how they set the `site` param for the `redirect_uri`. I assume it's the same as how we set it, but maybe they use the correct site param.

Reviewer: Did we confirm this?

Author: confirmed: users should use the site param for the token request, domain param for all other requests

   - The `domain` query parameter provides the [Datadog site][17] that the authorizing user is in, and will be needed to construct the URL for this POST request to the token endpoint: 

>        - For users authorizing from the US1, Fed1, and EU1 Datadog sites, the URL will be: `https://api.<domain>/oauth2/v1/token?...`

Author: @bgoldberg122 Just want to confirm that these are correct for the token endpoint

Reviewer: These are the [docs](https://docs.datadoghq.com/developers/authorization/oauth2_endpoints/?tab=tokenendpoints#exchange-authorization-code-for-access-token). If you want more formal confirmation I would ask AAA

Author: Waiting on confirmation for this

Author: Finalized this with API platform team and updated docs accordingly

>        - For users authorizing from the US1, Fed1, and EU1 Datadog sites, the URL will be: `https://api.<domain>/oauth2/v1/token?...`

Author: ```suggestion
       - For users authorizing from the US1, Fed1, and EU1 Datadog sites, the URL is: `https://api.<domain>/oauth2/v1/token?...`
```

>        - For users authorizing from any other Datadog site, the URL will be: `https://<domain>/oauth2/v1/token?...`

Author: ```suggestion
       - For users authorizing from any other Datadog site, the URL is: `https://<domain>/oauth2/v1/token?...`
```

6. Upon success, you receive your `access_token` and `refresh_token` in the response body. Your application should display a confirmation page with the following message: `You may now close this tab`.

7. Use the `access_token` to make calls to Datadog API endpoints by sending it in as a part of the authorization header of your request: ```headers = {"Authorization": "Bearer {}".format(access_token)}```.

>     - Note that API endpoints are different in each Datadog site. For example, if a user is in the EU region, the events endpoint is `https://api.datadoghq.eu/api/v1/events`, while for users in US1, the events endpoint is `https://api.datadoghq.com/api/v1/events`.

Author: ```suggestion
    - **Note***: API endpoints are different in each Datadog site. For example, if a user is in the EU region, the events endpoint is `https://api.datadoghq.eu/api/v1/events`, while for users in US1, the events endpoint is `https://api.datadoghq.com/api/v1/events`.
```

>     - **Note***: API endpoints are different in each Datadog site. For example, if a user is in the EU region, the events endpoint is `https://api.datadoghq.eu/api/v1/events`, while for users in US1, the events endpoint is `https://api.datadoghq.com/api/v1/events`.
>     - You can use the `domain` query parameter directly in these API calls to ensure the correct endpoint is being contacted. For example, to make a call to the events endpoint, you would build your request URL as `https://api.<domain>/api/v1/events`.

Reviewer 2: ```suggestion
    - **Note***: API endpoints are different in each Datadog site. For example, if a user is in the EU region, the Events endpoint is `https://api.datadoghq.eu/api/v1/events`, while for users in US1, the Events endpoint is `https://api.datadoghq.com/api/v1/events`.
    - You can use the `domain` query parameter directly in these API calls to ensure the correct endpoint is being contacted. For example, to make a call to the Events endpoint, you would build your request URL as `https://api.<domain>/api/v1/events`.
```

>     - Some endpoints may also require an API key, which is created in Step 8 below. 

Author: ```suggestion
    - Some endpoints may also require an API key, which is created in Step 8. 
```

### Initiate authorization from a third-party location 

> Users can start the authorization process in Datadog by clicking **Connect Accounts** in the integration tile. When a user clicks Connect Accounts in Datadog, information regarding their [Datadog site][17] is sent on redirect to the `onboarding_url`. The user's Datadog site is required to make API calls on behalf of the user and receive an authorization code. If a user initiates authorization from the _integration's external website_, the user's site information is not provided. 

Reviewer: We may want to use the word `domain` instead of `site` everywhere now to avoid confusion with the other url param

Author: I feel like officially it's called the Datadog site, while the domain param is specific to this OAuth implementation

### Initiate authorization from a third-party location 

> Users should start the authorization process in Datadog by clicking **Connect Accounts** in the integration tile. When a user clicks Connect Accounts in Datadog, information regarding their [Datadog site][17] is sent on redirect to the `onboarding_url`. The user's Datadog site is required in order to make API calls on behalf of the user and receive an authorization code. If a user initiates authorization from the _integration's external website_, the user's site information will not be provided. 

Reviewer 2: ```suggestion

Users can start the authorization process in Datadog by clicking **Connect Accounts** in the integration tile. When a user clicks Connect Accounts in Datadog, information regarding their [Datadog site][17] is sent on redirect to the `onboarding_url`. The user's Datadog site is required to make API calls on behalf of the user and receive an authorization code. If a user initiates authorization from the _integration's external website_, the user's site information is not provided. 
```

> Additionally, when users initiate authorization from the Datadog integration tile, they are required to have corresponding permissions for all requested scopes. If authorization is initiated from somewhere other than the integration tile, users without all of the required permissions may complete authorization (but are prompted to re-authorize with proper permissions when returning to the integration tile). 

Reviewer 2: ```suggestion
Additionally, when users initiate authorization from the Datadog integration tile, they are required to have corresponding permissions for all requested scopes. If authorization is initiated from somewhere other than the integration tile, users without all of the required permissions may complete authorization (but are prompted to re-authorize with proper permissions when they return to the integration tile). 
```

> Therefore, it is **strongly recommended** that partners prompt users to initiate authorization from Datadog, rather than from their own platform.

Reviewer 2: ```suggestion
Datadog recommends that partners prompt users to initiate authorization from Datadog, rather than from their own platform.
```

> While it is not recommended, if you would still like to support initiating authorization from a third-party location anywhere outside of the Datadog integration tile, ensure that you are able to support users in all Datadog sites, and are willing to continue to support new Datadog sites as they may be created. To ensure that users are authorizing in the correct site, always direct them to the US1 Datadog site (`app.datadoghq.com`), and from there, they can select their region.

Reviewer 2: ```suggestion
If you would like to support initiating authorization from a third-party location outside of the Datadog integration tile, ensure that you are able to support users in all Datadog sites, and are willing to continue to support new Datadog sites as they are created. To ensure that users are authorizing in the correct site, direct them to the US1 Datadog site (`app.datadoghq.com`), and from there, they can select their region.
```

Reviewer 2: ```suggestion
While Datadog does not recommend supporting initiating authorization from a third-party location anywhere outside of the Datadog integration tile, you must ensure that you are able to support users in all Datadog sites, and are willing to continue to support new Datadog sites as they may be created if you would like to. To ensure that users are authorizing in the correct site, always direct them to the US1 Datadog site (`app.datadoghq.com`), and from there, they can select their region.
```
